Reducing variation of tree creation methods.

paulp opened the pull request:
TreeDSL has no future - it was always a temporary measure
waiting for something like quasiquotes to come along. In this
commit I cull as much of it as I can, especially the delicate
matter of creating new DefDefs and ValDefs, which I completely
turn over to the old style creators.

I unified all the symbol-based DefDef and ValDef creators under
a single method, since it was yet another place where ctrl-C and
ctrl-V were being punched with glee. Added copyTypeDef creation
method to accompany the others.

Eliminated as many redundant positioning calls as I could find.
If you are creating a DefTree tree based on a symbol, it will
always have an atPos(sym.pos) { ... } wrapped around it. You
don't need another one.

All of this is motivated by positions work: positions are
assigned in so many places and in such an ad hoc fashion that
it is impossible to bring consistency to that without first
bringing some consistency to tree creation.

retronym commented on test/files/neg/t284.scala:
```diff
@@ -1,6 +1,5 @@
-  }
+  def f1(a: T): Unit { }
+  def f2(a: T): Unit
+  def f3(a: T) { }
```
This part looks good, as it wasn't originally intended as a test of brace healing. a626f6253829d0c8512551e8ae1290511cb077f1

paulp replied:
"authored 5 years ago", yikes. Yep, that was before I devised -Xfatal-warnings. In those days we made things fail by making them fail. I wore an onion on my belt, as was the style at the time.

retronym replied:
Still sounds better than my [5 years ago](https://jira.springsource.org/browse/SPR-4639) moment that I stumbled upon today!

paulp replied:
You're right! You can even one-up me on this level.

retronym replied:
In that case, it was more one-downsmapship :)

retronym commented on test/files/run/existential-rangepos.check:
```diff
@@ -1,6 +1,6 @@
 [[syntax trees at end of                    patmat]] // newSource1.scala
 [0:76]package [0:0]<empty> {
-  [0:76]abstract class A[[17:18]T[17:18]] extends [20:76][76]scala.AnyRef {
+  [0:76]abstract class A[[17:18]T[17]] extends [20:76][76]scala.AnyRef {
```
What part of the patch caused this change? Which tree node actually changed position? (I'm not 100% sure how to read these rendered range positions)

paulp replied:
The position on what should be the TypeDef in A's definition is focused [17] rather than a range position [17:18]. It makes zero difference in practice. I am not sure what the change was, but somewhere a TypeDef was probably getting sym.pos and is now getting sym.pos.focus.

paulp replied:
Actually I think it's the rhs of the TypeDef, that being a TypeTree.

paulp replied:
Okay, an empty TypeBoundsTree.

```
[17:18]TypeDef(
  <param>
  "T"
  []
  [18]TypeBoundsTree(
    <empty>
    <empty>
  )
)
```

retronym replied:
Okay, I think that this is fine. Creating a tree from just a symbol, as is done in tree printing post-parser, is inherently lossy wrt positions. So long as the approximated positions obey the range-pos invariants, we're fine.

paulp replied:
Too late, outdated diff discusser!

paulp commented on src/compiler/scala/tools/nsc/typechecker/MethodSynthesis.scala:
```diff
@@ -413,24 +392,16 @@ trait MethodSynthesis {
-            if (mods.isDeferred) EmptyTree else fieldSelection
-          ) setSymbol derivedSym
         }
+        tpt setPos tree.tpt.pos.focus
```
@retronym it's this line - I'm focusing the position always rather than selectively on a hard-to-justify subset of typetrees.

retronym replied:
I'm not sure this is the place.

```
% cat sandbox/test.scala
class A[T <: Int, U]

% git diff
diff --git a/src/compiler/scala/tools/nsc/typechecker/MethodSynthesis.scala b/src/compiler/scala/tools/nsc/typechecker/MethodSynthesis.scala
index 49f3a74..6be2861 100644
--- a/src/compiler/scala/tools/nsc/typechecker/MethodSynthesis.scala
+++ b/src/compiler/scala/tools/nsc/typechecker/MethodSynthesis.scala
@@ -399,6 +399,7 @@ trait MethodSynthesis {
           case _ if mods.isDeferred   => TypeTree() setOriginal tree.tpt // keep type tree of original abstract field
           case tp                     => TypeTree(tp)
         }
+        println((tpt, tree.tpt.pos, tree.tpt.pos.focus))
         tpt setPos tree.tpt.pos.focus
       }
       override def derivedTree: DefDef = newDefDef(derivedSym, derivedRhs)(tpt = derivedTpt)
% qbin/scalac -Yrangepos  sandbox/test.scala

% qbin/scalac -Xprint:parser,patmat -Yrangepos -Xprint-pos sandbox/test.scala > sandbox/new.log

% scalac-hash v2.11.0-M5 -Xprint:parser,patmat -Yrangepos -Xprint-pos sandbox/test.scala > sandbox/old.log
[info] v2.11.0-M5 => /Users/jason/usr/scala-v2.11.0-M5-0-gd6fe890

% cat sandbox/new.log | egrep '(trees|class)'
[[syntax trees at end of                    parser]] // test.scala
  [0:20]class A[[8:16]T[13:16] <: [13:16]Int, [18:19]U[19]] extends [20:20][20]scala.AnyRef {
[[syntax trees at end of                    patmat]] // test.scala
  [0:20]class A[[8:16]T[8] <: [8]Int, [18:19]U[18]] extends [20:20][20]scala.AnyRef {

% cat sandbox/old.log | egrep '(trees|class)'
[[syntax trees at end of                    parser]] // test.scala
  [0:20]class A[[8:16]T[13:16] <: [13:16]Int, [18:19]U[19]] extends [20:20][20]scala.AnyRef {
[[syntax trees at end of                    patmat]] // test.scala
  [0:20]class A[[8:16]T[8:16] <: [8]Int, [18:19]U[18:19]] extends [20:20][20]scala.AnyRef {
```

paulp replied:
A greater mystery I now face is if I compile this from the command line, before and after my patch:

```
abstract class A[T] {
  val foo: Set[_ <: T] = null
  val bar: Set[_ <: T]
}
```

with `scalac -Xprint:patmat -Yrangepos -Xprint-pos -Yshow-trees`
I get `<18>` as the position in both cases. The only difference in the tree comes later, at the expected point.

``` diff
         [-[63:74]<tpt>-]{+[63]<tpt>+} // tree.tpe=Set[_ <: T]
```

paulp replied:
That appears to be an artifact of -Yshow-trees. I get the 17 vs 17:18 diff without it, which means I will soon know why.

retronym replied:
I think the positions in my diff get printed from the symbols, rather than the trees.

retronym replied:
Somewhere in https://github.com/scala/scala/blob/master/src/compiler/scala/tools/nsc/ast/Printers.scala#L21

paulp commented on src/reflect/scala/reflect/internal/Trees.scala:
```diff
@@ -589,6 +589,10 @@ trait Trees extends api.Trees { self: SymbolTable =>
+  def TypeTree(sym: Symbol): TypeTree = atPos(sym.pos.focus)(TypeTree(sym.tpe_*.finalResultType))
+
+  def TypeBoundsTree(bounds: TypeBounds): TypeBoundsTree = TypeBoundsTree(TypeTree(bounds.lo), TypeTree(bounds.hi))
+  def TypeBoundsTree(sym: Symbol): TypeBoundsTree        = atPos(sym.pos.focus)(TypeBoundsTree(sym.info.bounds))
```
This spot.

retronym commented on src/reflect/scala/reflect/internal/Trees.scala:
```diff
@@ -1657,6 +1646,22 @@ trait Trees extends api.Trees { self: SymbolTable =>
+      )
+    case t =>
+      sys.error("Not a TypeDef: " + t + "/" + t.getClass)
+  }
```
This was added independently in the just-merged QuasiQuotes PR, so a rebase is needed: https://github.com/scala/scala/pull/2910/files#L17R1660

paulp replied:
Wow is that bit-for-bit identical? I guess it's not so surprising in this instance, but still.

paulp replied:
It was! It rebased without conflict despite independent origins.

xeno-by commented on test/files/run/macro-typecheck-macrosdisabled2.check:
```diff
@@ -1,5 +1,5 @@
 {
-  val $u: reflect.runtime.universe.type = scala.reflect.runtime.`package`.universe;
+  val $u: scala.reflect.api.JavaUniverse = scala.reflect.runtime.`package`.universe;
```
I wonder why this happens

paulp replied:
.finalResultType. I considered it an improvement.

paulp replied:
...because things don't generally get typed as their own singleton type.

paulp replied:
Scratch that last comment. Maybe it's still an improvement, I'd have to look.

paulp replied:
I can't tell if it makes any sense by looking at the code in the test.

paulp replied:
But to clarify, it's because I'm doing this uniformly:

```
def TypeTree(sym: Symbol): TypeTree = atPos(sym.pos.focus)(TypeTree(sym.tpe_*.finalResultType))
```

Whereas it used to do things differently based on whether it was a MethodSymbol or something like that.

xeno-by replied:
In this case an explicit singleton type is desireable, so that $u.Expr has the correct type.

paulp replied:
But are you relying on what-the-implementation-happened-to-do for the singleton type? I think if it were explicitly given a singleton type it would keep it, else everything else wouldn't work.

paulp replied:
In case that is not clear - to be affected by that line of code you must be creating a TypeTree from a symbol. When you create a TypeTree from a symbol, you pretty much turn your fate over to the gods, because a symbol isn't enough to tell us what a type is. It is no surprise that with all the different ways of creating things, sometimes you get what you want, and doing things consistently always means having to chase down those works-by-accident situations and redoing them in a way which will hold up. My intuition is that this is one of those cases.

retronym replied:
It is likely that the conversion of a `Symbol` to `TypeTree` is happening during tree printing in the code below. I think we need to restore the previous behaviour.

``` scala
    override def print(args: Any*): Unit = args foreach {
      case tree: Tree =>
        printPosition(tree)
        printTree(
            if (tree.isDef && tree.symbol != NoSymbol && tree.symbol.isInitialized) {
              tree match {
                case ClassDef(_, _, _, impl @ Template(ps, emptyValDef, body))
                if (tree.symbol.thisSym != tree.symbol) =>
                  ClassDef(tree.symbol, Template(ps, ValDef(tree.symbol.thisSym), body))
                case ClassDef(_, _, _, impl)           => ClassDef(tree.symbol, impl)
                case ModuleDef(_, _, impl)             => ModuleDef(tree.symbol, impl)
                case ValDef(_, _, _, rhs)              => ValDef(tree.symbol, rhs)
```

retronym replied:
Although perhaps a `SingletonType` ought to override `finalResultType` to just return `this`, rather than inheriting:

``` scala
// SimpleTypeProxy
override def finalResultType = underlying.finalResultType
```

paulp replied:
Now you're speaking my language.

paulp replied:
This being the right thing to do is reinforced by the fact that resultType does NOT appear in SimpleTypeProxy, so SingletonType inherits the identity implementation in that case.

retronym replied:
Watch out, you'll be crossing paths with [MCFTIDE](https://github.com/scala/scala/commit/a205b6b06#L40R128)

paulp replied:
![neo](https://f.cloud.github.com/assets/9135/1122475/8d4e49cc-1adc-11e3-96b3-b1be69e068b4.jpg)

retronym replied:
I'll bet that contributes to unwanted widening in extractor calls in pattern matching, too.

``` scala
class ExtractorCallProd(
      ...
      // to which type should the previous binder be casted?
      def paramType  = constructorTp.finalResultType
      def resultType = fun.tpe.finalResultType
```

I was testing out variations of your pattern matcher patch with:

``` scala
object Test {
  // Has never worked, but seems desirable given the recent changes to
  // pattern type inference.
  val a = ""
  object Id {
    def unapply(xxxx: Any): Some[a.type] = Some[a.type](a)
  }
  val b: a.type = (a: a.type) match {
    case Id(x) => x
  }
}
```

paulp replied:
You are right, that now compiles.

paulp replied:
These are both x.type in my forthcoming push -f. Tests are pretty deep, cross fingers.

```
// Current trunk though
scala> val x = NoSymbol.newTermSymbol("x") setInfo IntClass.tpe
x: $r.intp.global.TermSymbol = value x

scala> val tp = MethodType(Nil, SingleType(x.owner.thisType, x))
tp: $r.intp.global.MethodType = ()x.type

scala> println ((tp.resultType, tp.finalResultType))
(x.type,Int)
```

paulp committed:
Corrects behavior of finalResultType.

The implementation had come to depend on finalResultType
accidentally doing things beyond its charter - in particular,
widening types. After hunting down and fixing the call sites
depending on the bugs, I was able to rewrite the method to do
only what it's supposed to do.

I threw in a different way of writing it entirely to suggest how
some correctness might be obtained in the future. It's a lot
harder for a method written like this to break.

paulp commented:
I rebased this on top of #2934, which is the runaway winner in the world of final result types.

paulp committed:
Widen type of outer accessor.

This is the key ingredient so TypeTree(sym) can resist
widening the type.

paulp committed:
Reducing variation of tree creation methods.

TreeDSL has no future - it was always a temporary measure
waiting for something like quasiquotes to come along. In this
commit I cull as much of it as I can, especially the delicate
matter of creating new DefDefs and ValDefs, which I completely
turn over to the old style creators.

I unified all the symbol-based DefDef and ValDef creators under
a single method, since it was yet another place where ctrl-C and
ctrl-V were being punched with glee. Was beaten to the punch on
adding copyTypeDef to fill out the *Def creators.

Eliminated as many redundant positioning calls as I could find.
If you are creating a DefTree tree based on a symbol, it will
always have an atPos(sym.pos) { ... } wrapped around it. You
don't need another one.

All of this is motivated by positions work: positions are
assigned in so many places and in such an ad hoc fashion that
it is impossible to bring consistency to that without first
bringing some consistency to tree creation.

paulp commented:
@retronym Can I get a stamp of some kind for this? I know it's the weekend and everything, but monday is my "last" day and then I'm working for free. I noticed it's also your birthday, it's so fitting.

retronym commented:
Stamp pending in about 12 h.

retronym commented:
I've taken a look at the [diffs of the trees](https://gist.github.com/retronym/6569792). 

## ConstantType

Constant folding into the RHS of accessors indeed is working again. As you said, the accessor now carries the `ConstantType`. I believe that this lies somewhere between an improvement and benign.

```
     final private[this] val ASSERTION: Int(2000) = 2000;
-    final <stable> <accessor> def ASSERTION: Int = 2000;
+    final <stable> <accessor> def ASSERTION: Int(2000) = 2000;
```

I checked if ScalaDoc would differ before/after, but it seems like it already was showing the constant type.

```
% RUNNER=scaladoc scala-hash v2.11.0-M5 sandbox/const.scala
[info] v2.11.0-M5 => /Users/jason/usr/scala-v2.11.0-M5-0-gd6fe890
warning: dropping dependency on node with no phase object: pickler
model contains 2 documentable templates
```

![image](https://f.cloud.github.com/assets/65551/1145171/2db5c2e2-1df7-11e3-93a7-94a4ab707aba.png)

```
% build/pack/bin/scaladoc sandbox/const.scala
model contains 2 documentable templates
```

![image](https://f.cloud.github.com/assets/65551/1145173/530b134e-1df7-11e3-846e-5b8ca16026a0.png)

If we were being paranoid we could `deconst` accessor types to restore the status quo. But I think we can merge it as is.

## ThisType

The next difference relates to `ThisType`

```
# BEFORE
% egrep '(syntax tree|override def buffered)' sandbox/log-v2.11.0-M5-93-g33a819f.txt
[[syntax trees at end of                    parser]] // AnyVal.scala
      override def buffered: this.type = this
[[syntax trees at end of                     namer]] // AnyVal.scala
      override def buffered: this.type = this
[[syntax trees at end of            packageobjects]] // AnyVal.scala
      override def buffered: this.type = this
[[syntax trees at end of                     typer]] // AnyVal.scala
      override def buffered: scala.collection.BufferedIterator[A] = this
```

```
# AFTER
% egrep '(syntax tree|override def buffered)' sandbox/log-2931-330ead53be956.txt
[[syntax trees at end of                    parser]] // AnyVal.scala
      override def buffered: this.type = this
[[syntax trees at end of                     namer]] // AnyVal.scala
      override def buffered: this.type = this
[[syntax trees at end of            packageobjects]] // AnyVal.scala
      override def buffered: this.type = this
[[syntax trees at end of                     typer]] // AnyVal.scala
      override def buffered: BufferedIterator.this.type = this
```

Hmm, what was going on before hand. Seen in isolation:

```
% cat sandbox/a.scala
trait Temp {
  def buffered: this.type = this
}

% scalac-hash v2.11.0-M5 -Xprint:parser,typer sandbox/a.scala
[info] v2.11.0-M5 => /Users/jason/usr/scala-v2.11.0-M5-0-gd6fe890
[[syntax trees at end of                    parser]] // a.scala
package <empty> {
  abstract trait Temp extends scala.AnyRef {
    def $init$() = {
      ()
    };
    def buffered: this.type = this
  }
}

[[syntax trees at end of                     typer]] // a.scala
package <empty> {
  abstract trait Temp extends scala.AnyRef {
    def /*Temp*/$init$(): Unit = {
      ()
    };
    def buffered: Temp = this
  }
}

% scala-hash v2.11.0-M5
[info] v2.11.0-M5 => /Users/jason/usr/scala-v2.11.0-M5-0-gd6fe890
Welcome to Scala version 2.11.0-20130906-140349-d6fe8905af (Java HotSpot(TM) 64-Bit Server VM, Java 1.6.0_37).
Type in expressions to have them evaluated.
Type :help for more information.

scala> :power
** Power User mode enabled - BEEP WHIR GYVE **
** :phase has been set to 'typer'.          **
** scala.tools.nsc._ has been imported      **
** global._, definitions._ also imported    **
** Try  :help, :vals, power.<tab>           **

scala> typeOf[Temp].member(TermName("buffered")).info
res0: $r.intp.global.Type = => Temp.this.type

scala> exitingTyper(typeOf[Temp].member(TermName("buffered")).info)
res1: $r.intp.global.Type = => Temp.this.type
```

Oh, I see. `TreePrinters` calls `DefDef(sym)` and gets the unwanted widening of the result type. The `:power` mode session at the end shows that the actual pickled symbol was correct. Now we can see that in high-fidelity tree printing. Win!

## `scala.` prefixes.

They appear in some places:

```
     case <synthetic> def unapply[T1, T2, T3](x$0: (T1, T2, T3)): Option[(T1, T2, T3)] = if (x$0.==(null))
       scala.this.None
     else
-      Some.apply[(T1, T2, T3)](Tuple3.apply[T1, T2, T3](x$0._1, x$0._2, x$0._3));
+      Some.apply[(T1, T2, T3)](scala.Tuple3.apply[T1, T2, T3](x$0._1, x$0._2, x$0._3));
```

And disappear in others:

```
     private[this] val Right: util.Right.type = scala.util.Right;
-    <stable> <accessor> def Right(): scala.util.Right.type = `package`.this.Right
+    <stable> <accessor> def Right(): util.Right.type = `package`.this.Right
```

I'd like to understand what's responsible for this change, but it looks benign, so I won't hold up this PR.

retronym commented:
LGTM.
